dataactvalidator/validation_handlers/validationManager.py

```diff
@@ -318,16 +320,45 @@ def runCrossValidation(self, jobId, interfaces):
         failures.extend(Validator.crossValidateSql(rules, submissionId))
         bucketName = CONFIG_BROKER['aws_bucket']
         regionName = CONFIG_BROKER['aws_region']
-        errorFileName = self.getFileName(interfaces.jobDb.getCrossFileReportPath(submissionId))
         errorDb = interfaces.errorDb
 
-        with self.getWriter(regionName, bucketName, errorFileName,
-                            self.crossFileReportHeaders) as writer:
-            for failure in failures:
-                writer.write(failure)
-                errorDb.recordRowError(jobId, "cross_file",
-                    failure[0], failure[2], failure[4], failure[5])
-            writer.finishBatch()
+        # Get list of file types
+        fileTypes = interfaces.validationDb.getFileTypeList()
+        # Create dict mapping file types to empty lists
+        fileTypeDict = {targetFile:[] for targetFile in fileTypes}
+
+        # Sort failures by source and target file
+        failureDict = {sourceFile:copy.deepcopy(fileTypeDict) for sourceFile in fileTypes}
+
+        for failure in failures:
+            errorDb.recordRowError(jobId, "cross_file",
+                    failure[0], failure[2], failure[4], failure[5], failure[6], failure[7])
+            # Put failure into list for this source and target file
+            sourceType = interfaces.validationDb.getFileTypeById(failure[6])
+            targetType = interfaces.validationDb.getFileTypeById(failure[7])
+            if failure[7] < failure[6]:
+                # Lower ID goes first
+                failureDict[targetType][sourceType].append(failure)
+            elif failure[7] == failure[6]:
+                raise ValueError("Both files in cross file are the same type")
+            else:
+                failureDict[sourceType][targetType].append(failure)
+
+        # For each pair of files create an error report
+        for sourceFile in fileTypes:
```

Comment:
Target file has now been added, hopefully this makes it a bit clearer that these two errors are happening in opposite directions.